src/gc.c

```diff
@@ -2115,6 +2115,39 @@ JL_DLLEXPORT void *jl_gc_managed_realloc(void *d, size_t sz, size_t oldsz,
     return b;
 }
 
+JL_DLLEXPORT jl_value_t *jl_gc_realloc_string(jl_value_t *s, size_t sz)
+{
+    size_t len = jl_string_len(s);
+    if (sz <= len) return s;
+    size_t strsz = len + sizeof(size_t) + 1;
+    if (strsz <= GC_MAX_SZCLASS) {
+        // pool allocated; can't be grown in place so allocate a new object.
+        jl_value_t *snew = jl_alloc_string(sz);
+        memcpy(jl_string_data(snew), jl_string_data(s), len);
+        return snew;
+    }
+    size_t newsz = sz + sizeof(size_t) + 1;
+    size_t offs = offsetof(bigval_t, header);
+    size_t allocsz = LLT_ALIGN(newsz + offs, JL_CACHE_BYTE_ALIGNMENT);
+    if (allocsz < sz)  // overflow in adding offs, size was "negative"
+        jl_throw(jl_memory_exception);
+    bigval_t *hdr = bigval_header(jl_astaggedvalue(s));
+    gc_big_object_unlink(hdr);
```

Comment:
I believe this is only safe for young objects. If we do this for old objects it should not be linked back to the young list and linking it to the old list can hit https://github.com/JuliaLang/julia/issues/17971. My WIP parallel gc branch already fixes that problem but won't be ready for 0.6. Can we conditionally do the realloc for young objects and possible handle old ones later?